dockstore-webservice/src/main/java/io/dockstore/webservice/resources/ServiceResource.java

```diff
@@ -0,0 +1,227 @@
+import io.dockstore.webservice.jdbi.UserDAO;
+import io.dockstore.webservice.jdbi.WorkflowDAO;
+import io.dropwizard.auth.Auth;
+import io.dropwizard.hibernate.UnitOfWork;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.Authorization;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.HttpClient;
+import org.hibernate.SessionFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static io.dockstore.webservice.Constants.JWT_SECURITY_DEFINITION_NAME;
+import static io.dockstore.webservice.Constants.LAMBDA_FAILURE;
+import static io.dockstore.webservice.core.WorkflowMode.SERVICE;
+
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+@Path("/workflows")
+@Produces(MediaType.APPLICATION_JSON)
+public class ServiceResource extends AbstractWorkflowResource {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ServiceResource.class);
+    private final WorkflowDAO workflowDAO;
+    private final UserDAO userDAO;
+    private final TokenDAO tokenDAO;
+    private final String gitHubPrivateKeyFile;
+    private final String gitHubAppId;
+
+    public ServiceResource(HttpClient client, SessionFactory sessionFactory, DockstoreWebserviceConfiguration configuration) {
+        super(client, sessionFactory, configuration);
+        this.workflowDAO = new WorkflowDAO(sessionFactory);
+        this.userDAO = new UserDAO(sessionFactory);
+        this.tokenDAO = new TokenDAO(sessionFactory);
+        gitHubAppId = configuration.getGitHubAppId();
+        gitHubPrivateKeyFile = configuration.getGitHubAppPrivateKeyFile();
+    }
+
+    @POST
+    @Path("/path/service/upsertVersion/")
+    @Timed
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @UnitOfWork
+    @RolesAllowed({ "curator", "admin" })
+    @ApiOperation(value = "Add or update a service version for a given GitHub tag for a service with the given repository (ex. dockstore/dockstore-ui2).", notes = "To be called by a lambda function. Error code 418 is returned to tell lambda not to retry.", authorizations = {
+            @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, response = Workflow.class)
+    public Workflow upsertServiceVersion(@ApiParam(hidden = true) @Auth User user,
+            @ApiParam(value = "Repository path", required = true) @FormParam("repository") String repository,
+            @ApiParam(value = "Name of user on GitHub", required = true) @FormParam("username") String username,
+            @ApiParam(value = "Git reference for new GitHub tag", required = true) @FormParam("gitReference") String gitReference,
+            @ApiParam(value = "GitHub installation ID", required = true) @FormParam("installationId") String installationId) {
+
+        // Retrieve the user who triggered the call (may not exist on Dockstore)
+        User sendingUser = GitHubHelper.findUserByGitHubUsername(this.tokenDAO, this.userDAO, username, false);
```

Comment:
I wonder if enough has actually been moved into the abstract class. 
I feel like a workflow equivalent of upsertVersion would do fairly similar things like finding the user, getting a token, adding the user, etc.

Reply:
You are right that more needs to be done. I approached this PR more as a scaffolding for further abstraction rather than a full-blown implementation. Given we are approaching feature complete, I didn't want to risk breaking anything, especially any workflow code.

But on the other hand the services code is new, so it should be safe to abstract further.

Reply:
I'm going ahead with further refactoring.